### PR TITLE
refactor(oas): extract URI-template logic

### PR DESCRIPTION
**Pull request by pmstss**
relates-to #120

### CODE REVIEW

**Comment by pmstss** [comment]
must be rebased after #134 merge

**Comment by pmstss** [comment]
@derevnjuk complexity complains are caused by `PostDataConverter`, which refactoring we were decided to postpone;
is ready for review, but must be rebased after https://github.com/NeuraLegion/har-sdk/pull/134 merge

File: packages/oas/src/converter/parts/SubConvertersFactory.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import {
+  Header,
+  OpenAPI,
+  OpenAPIV2,
+  OpenAPIV3,
+  QueryString
+} from '@har-sdk/core';
+
+export class SubConvertersFactory {
+  private subConverters = new Map<SubPart, SubConverter<unknown>>();
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/SubConvertersFactory.ts:18]:
```suggestion
  private readonly subConverters = new Map<SubPart, SubConverter<unknown>>();
```

File: packages/oas/src/converter/parts/SubConvertersFactory.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    private readonly sampler: Sampler
+  ) {}
+
+  public get(type: SubPart): SubConverter<unknown> {
+    if (!this.subConverters.has(type)) {
+      this.subConverters.set(type, this.createSubConverter(type));
+    }
+
+    return this.subConverters.get(type);
+  }
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/SubConvertersFactory.ts:31]: It looks like a registry, but not a factory) Let's rename accordinaly

**Review by derevnjuk — COMMENTED** [review]
Let's rebase it first, and then I can continue the review. Currently, it appears to be not an easy task)

File: packages/oas/src/converter/parts/SubConvertersRegistry.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  private createSubConverter(type: SubPart): SubConverter<any> {
+    switch (type) {
+      case SubPart.HEADERS:
+        return this.createHeadersConverter();
+      case SubPart.PATH:
+        return this.createPathConverter();
+      case SubPart.POST_DATA:
+        return new PostDataConverter(this.spec, this.sampler);
+      case SubPart.QUERY_STRING:
+        return this.createQueryStringConverter();
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/SubConvertersRegistry.ts:42]:
```suggestion
      case SubPart.QUERY_STRING:
        return this.createQueryStringConverter();
      default:
        throw new TypeError(`${type} subconverter is not supported`)
```

File: packages/oas/src/converter/parts/headers/HeadersConverter.ts [hunk 4]
```diff
@@ -35,31 +45,53 @@ export class HeadersConverter implements SubConverter<Header[]> {
...
+  protected createApiKeyHeader(scheme: SecuritySchemeApiKey): Header {
+    return this.createHeader(scheme.name.toLowerCase(), 'REPLACE_KEY_VALUE');
+  }
+
+  protected createHeader(name: string, value: string): Header {
+    return {
+      name,
+      value
+    };
+  }
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/headers/HeadersConverter.ts:73]:
```suggestion
  protected createApiKeyHeader(scheme: SecuritySchemeApiKey): Header {
    return this.createHeader(scheme.name, 'REPLACE_KEY_VALUE');
  }

  protected createHeader(name: string, value: string): Header {
    return {
      value,
      name: name.toLowerCase()
    };
  }
```

File: packages/oas/src/converter/parts/headers/HeadersConverter.ts [hunk 5]
```diff
@@ -35,31 +45,53 @@ export class HeadersConverter implements SubConverter<Header[]> {
...
+  protected parseApiKeyScheme(
+    securityScheme: SecuritySchemeObject
+  ): Header | undefined {
+    if ('in' in securityScheme && securityScheme.in === 'header') {
+      return this.createApiKeyHeader(securityScheme);
     }
+  }
+
+  protected createBearerAuthHeader(): Header {
+    return this.createHeader('authorization', 'Bearer REPLACE_BEARER_TOKEN');
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/headers/HeadersConverter.ts:57]: Why don't use the sampler here?

```suggestion
    const token = this.sampler.sample({ type: 'string', maxLength: 10, minLength: 5 });
    
    return this.createHeader('authorization', `Bearer ${token}`);
```

File: packages/oas/src/converter/parts/headers/HeadersConverter.ts [hunk 6]
```diff
@@ -35,31 +45,53 @@ export class HeadersConverter implements SubConverter<Header[]> {
...
+  }
 
-    return this.createHeaders('content-type', values);
+  protected createBasicAuthHeader(): Header {
+    return this.createHeader('authorization', 'Basic REPLACE_BASIC_AUTH');
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/headers/HeadersConverter.ts:61]: ditto

File: packages/oas/src/converter/parts/headers/HeadersConverter.ts [hunk 7]
```diff
@@ -35,31 +45,53 @@ export class HeadersConverter implements SubConverter<Header[]> {
...
   }
 
-  private createAcceptHeaders(pathObj: OperationObject): Header[] {
-    return this.createHeaders(
-      'accept',
-      'produces' in pathObj && Array.isArray(pathObj.produces)
-        ? pathObj.produces
-        : []
+  protected createApiKeyHeader(scheme: SecuritySchemeApiKey): Header {
+    return this.createHeader(scheme.name.toLowerCase(), 'REPLACE_KEY_VALUE');
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/headers/HeadersConverter.ts:65]: ditto

File: packages/oas/src/converter/parts/headers/Oas3HeadersConverter.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+export class Oas3HeadersConverter extends HeadersConverter<OpenAPIV3.Document> {
+  constructor(spec: OpenAPIV3.Document, sampler: Sampler) {
+    super(spec, sampler);
+  }
+
+  protected createContentTypeHeaders(
+    pathObj: OpenAPIV3.OperationObject
+  ): Header[] {
+    const requestBody = pathObj.requestBody as OpenAPIV3.RequestBodyObject;
+    if (requestBody?.content != null && isObject(requestBody?.content)) {
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/headers/Oas3HeadersConverter.ts:15]:
```suggestion
    if (requestBody?.content && isObject(requestBody?.content)) {
```

File: packages/oas/src/converter/parts/path/Oas2PathConverter.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import { Sampler } from '../Sampler';
+import { Oas2ValueSerializer } from '../Oas2ValueSerializer';
+import { PathConverter } from './PathConverter';
+import { OpenAPIV2 } from '@har-sdk/core';
+
+export class Oas2PathConverter extends PathConverter<OpenAPIV2.Parameter> {
+  private oas2ValueSerializer = new Oas2ValueSerializer();
```
Comment (derevnjuk) [inline packages/oas/src/converter/parts/path/Oas2PathConverter.ts:7]:
```suggestion
  private readonly oas2ValueSerializer = new Oas2ValueSerializer();
```

**Review by derevnjuk — CHANGES_REQUESTED** [review]

**Review by derevnjuk — APPROVED** [review]

**Pull request merged**